### PR TITLE
Demo: Plotly plugin for openmct

### PR DESCRIPTION
**Pull request by JoelMcKinnon**
Created plotly plugin to compare with current functionality of overlay plot.

Allows addition of one or more telemetry objects. Currently hardcoded to UTC time axis and only allows multiple objects of same type to keep y axis consistent.

Can plot data in fixed timespan and local clock.

Testing notes:

1. Run npm install, npm start
2. Create multiple sign wave generators with different properties.
3. Create a Plotly Plot.
4. Drag in the SWGs you created earlier
5. Experiment with Time Conductor
6. Change to local clock
7. Experiment with start and end deltas
8. Save and reload.
9. Remove one or more telemetry objects using Elements pane.
10. Save and reload.

**Author Checklist**

- Changes address original issue? Yes
- Unit tests included and/or updated with changes? No
- Command line build passes? Yes
- Changes have been smoke-tested? Yes

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -5,6 +5,7 @@ jobs:
         - image: circleci/node:13-browsers
           environment:
             CHROME_BIN: "/usr/bin/google-chrome"
+        - resource_class: medium+
```
Comment (JoelMcKinnon) [inline .circleci/config.yml]: That's all it needed I guess. I guess I needed to look through the CircleCI docs instead of SO.

**Comment by akhenry** [comment]
I did some investigation of the circle-ci failures. I believe that what is happening is that chrome-headless is silently crashing when this runs on the circle-ci servers. Although we are able to bump the node heap space up to 4GB, the docker image that it runs on in CircleCI is limited to 4GB of memory in the free tier. So, I suspect we were close to the limit already, and Plotly has pushed us over the edge.

The problem occurs the moment Plotly is included from any of our source files (it's not enough to just include it from package.json because it won't be picked up by webpack).

We don't need to merge this any time soon, as this is just an experiment, so it's not a pressing issue. But in terms of pathways forward, I think there are a couple of options. We should look into at why our tests use so much memory in comparison to regular runtime usage. There may be ways of configuring or using our test tools that will reduce memory usage and test time. eg. https://github.com/webpack-contrib/karma-webpack/issues/281

It is also apparently possible to selectively import the parts of Plotly that we need without needing to import the whole shebang.

File: .circleci/config.yml (hunk 1 above)
Comment (akhenry) [reply .circleci/config.yml]: We can't use medium+ anyway as it transpires, it's not available on the free tier.

File: src/plugins/plotlyPlot/components/PlotlyViewLayout.vue [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+            delete this.subscriptions[keyString];
+        },
+        refreshData(bounds, isTick) {
+            this.bounds = bounds;
+
+            this.telemetryObjects.forEach((telemetryObject, index) => {
+                if(!isTick) {
+                    this.requestHistory(telemetryObject, index, false);
+                } else {
+                    if (this.timeRange === 0 || this.timeRange !== this.bounds.end - this.bounds.start) {
```
[the remark this answers is not in the document]
Comment (JoelMcKinnon) [reply src/plugins/plotlyPlot/components/PlotlyViewLayout.vue]: Fixed.

File: src/plugins/plotlyPlot/components/PlotlyViewLayout.vue [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+                    return;
+                }
+                this.updateData(datum, index);
+            });
+        },
+        unsubscribe(keyString) {
+            this.subscriptions[keyString]();
+            delete this.subscriptions[keyString];
+        },
+        refreshData(bounds, isTick) {
```
[the remark this answers is not in the document]
Comment (JoelMcKinnon) [reply src/plugins/plotlyPlot/components/PlotlyViewLayout.vue]: Done

File: src/plugins/plotlyPlot/components/PlotlyViewLayout.vue [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+            if (telemetryObject.type === 'generator') {
+                this.yAxisProp = 'sin';
+            } else if (telemetryObject.type === 'example.state-generator') {
+                this.yAxisProp = 'state';
+            } else if (telemetryObject.type === 'conditionSet') {
+                this.yAxisProp = 'output';
+            }
+        },
+        formatDatumX(datum) {
+            let timestamp = moment.utc(datum.utc).format('YYYY-MM-DDTHH:mm:ss[Z]');
```
[the remark this answers is not in the document]
Comment (JoelMcKinnon) [reply src/plugins/plotlyPlot/components/PlotlyViewLayout.vue]: Done

File: src/plugins/plotlyPlot/components/PlotlyViewLayout.vue [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+        getYAxisLabel(telemetryObject) {
+            this.setYAxisProp(telemetryObject);
+            const valueMetadatas = this.openmct.telemetry.getMetadata(telemetryObject).values();
+            const index = valueMetadatas.findIndex(value => value.key === this.yAxisProp);
+            const yLabel = valueMetadatas[index].name;
+
+            return yLabel;
+        },
+        setYAxisProp(telemetryObject) {
+            if (telemetryObject.type === 'generator') {
```
[the remark this answers is not in the document]
Comment (JoelMcKinnon) [reply src/plugins/plotlyPlot/components/PlotlyViewLayout.vue]: Done

**Comment by deeptailor** [comment]
@JoelMcKinnon - Took a quick look at the plotly work. It looks great, but I have a few suggestions:

1. The reason for using bounded rows is to keep track of the data, without having to do it separately in your plugin. It looks like bounded rows are keeping track of data and so is the plotly plugin. You should leverage the logic built into bounded rows, ie what andrew said above:

   `boundedTelemetryCollection.on('add', addToTrace);
    boundedTelemetryCollection.on('remove', removeFromTrace);`

This way you dont have to listen to bounds through your plugin, BoundedRows will do that for you.

**Comment by JoelMcKinnon** [comment]
> 1. The reason for using bounded rows is to keep track of the data, without having to do it separately in your plugin. It looks like bounded rows are keeping track of data and so is the plotly plugin. You should leverage the logic built into bounded rows, ie what andrew said above:
>    `boundedTelemetryCollection.on('add', addToTrace); boundedTelemetryCollection.on('remove', removeFromTrace);`

I don't see how my plugin is keeping track of data. It's using plotly to extend the traces when new data is received (in updateData) and replotting without data that is out of range in updatePlotRange(). It seems to me that boundedTelemetryRowCollection is adding a lot of unneeded complexity, but I must be missing something.

**Comment by JoelMcKinnon** [comment]
> 1. You are only keeping track of one set of meta data, what happens when you add multiple items, eg sine wave and state generator, that have different metadata?

I'm defining the keys for the labels and values within addTelemetryObject and using the object passed rather than the domain object. I know there would be a problem when adding different types of telemetry which is why I'm only testing with Sine Wave Generators for now, but as long as it's just a single type I can add as many as I want.

**Pull request closed without merging**